// @expo/build-tools/dist/platforms/android/generic/builder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const turtle_spawn_1 = __importDefault(require("@expo/turtle-spawn"));
const project_1 = require("../../../utils/project");
const credentials_1 = require("../utils/credentials");
const builder = async function buildGeneric(ctx) {
    var _a;
    const { job, logger } = ctx;
    const buildDirectory = await project_1.setup(ctx);
    const reactNativeProjectDirectory = path_1.default.join(buildDirectory, job.projectRootDirectory);
    const nativeProjectDirectory = path_1.default.join(reactNativeProjectDirectory, 'android');
    logger.info("Writing secrets to the project's directory");
    if (job.secrets.buildCredentials) {
        await credentials_1.restoreCredentials(reactNativeProjectDirectory, job);
    }
    logger.info('Starting gradle build');
    logger.debug(`Running './gradlew ${job.gradleCommand}' in ${nativeProjectDirectory}`);
    await turtle_spawn_1.default('./gradlew', [job.gradleCommand], {
        cwd: nativeProjectDirectory,
        logger,
        lineTransformer: (line) => {
            if (!line || /^\.+$/.exec(line)) {
                return null;
            }
            else {
                return line;
            }
        },
        env: {
            ...process.env,
            ...(_a = job.secrets) === null || _a === void 0 ? void 0 : _a.secretEnvs,
        }
    });
    return {
        projectDirectory: reactNativeProjectDirectory,
        artifactPath: job.artifactPath,
    };
};
exports.default = builder;
